.github/workflows/test.yml

```diff
@@ -0,0 +1,79 @@
+name: CI
+
+on: [push, pull_request]
+
+jobs:
+  render:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - uses: actions/setup-python@v4
+        with:
+          python-version: '3.11'
+
+      - run: python -m pip install cookiecutter
+      - run: python -m cookiecutter . --no-input
+
+      - uses: actions/upload-artifact@v3
+        with:
+          name: rendered-project
+          path: wagtail-llama-save
+
+  lint:
+    needs: render
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - uses: actions/setup-python@v4
+        with:
+          python-version: '3.11'
+
+      - uses: actions/download-artifact@v3
+        with:
+          name: rendered-project
+          path: /tmp/wagtail-llama-save
+
+      - working-directory: /tmp/wagtail-llama-save
+        run: |
+          git init
+          git add .
+          python -m pip install pre-commit
+          pre-commit run --all-files --show-diff-on-failure --color=always
```

Comment:
Again, I'm going with the approach of doubling down on pre-commit. That said, this bit probably needs some explanation.

What we want to do here is just run lint checks on the rendered output. We can't run pre-commit in the repo working directory though. Even if we invoke pre-commit from in a subdirectory, it tries to lint the whole repo including the raw templates (which are syntactically valid jinja templates, but aren't necessarily syntactically valid toml, json, python, etc). So instead of restoring the rendered project inside the working dir (like with all the other jobs), I'm restoring the rendered artefact to a location outside the template repo checkout. Then, in order to get pre-commit to see any of the files, we have to initialize a git repo and at least stage (if not commit) the files we want to check. Otherwise pre-commit will stash everything and tell us there are no files to check.

The need to jump through these hoops also prevents us from using `pre-commit/action`.

Of course, it is entirely possible I am being a massive pre-commit n00b. Please feel free to propose a better solution. If we need to roll with this I will add a comments to explain this.

Side note: This actually raises another source of friction I've found with pre-commit but had managed to block out of my mind:

If you completely delegate linter management to pre-commit and don't have your linters installed in your local environment, I haven't found any way to get pre-commit to run on an unstaged file. This makes sense in that pre-commit's design goal is to be a git hook manager, not a linter manager but if you don't have your linters installed in your local environment, I think that then means there is no way to run any of your lint checks until you've staged your changes for commit, and that is too late for me. Do you have any advice on a coping mechanism for this? Maybe I'm holding it wrong.

In the spirit of doubling down and delegating all linting concerns to pre-commit, I am trying _really hard_ to understand and like this tool. I still have some hard work ahead of me :laughing:

Reply:
`pre-commit run --all-files` will do it on all without the need to stage anything
`pre-commit run LINTER --all-files` will do so with a specific one defined in config
`pre-commit run --files file1 file2` for specific files

Reply:
That is what I thought/expected too, but it isn't what actually happens.

I have pushed 2 commits to demonstrate.

In 7b9bdf6 I have introduced some code style errors to make the build fail. If we look at the workflow run for that commit, the lint checks fail, as we would expect: https://github.com/wagtail/cookiecutter-wagtail-package/actions/runs/3893274743/jobs/6645681624

Then in 95484c3 I have removed the `git add .` command. The code style errors are still there - I didn't roll them back, but now the lint checks pass: https://github.com/wagtail/cookiecutter-wagtail-package/actions/runs/3893294131/jobs/6645723896
pre-commit reports it has "no files to check".

Note we _are_ running with the `--all-files` flag: https://github.com/wagtail/cookiecutter-wagtail-package/blob/95484c369aef5541949f6bec360ae056dc023807/.github/workflows/test.yml#L40

Reply:
hmm.. `--files FILE1 FILE2` works like a charm. Looking at https://github.com/pre-commit/pre-commit/issues/848 it seels `--all-files` does indeed only check staged files :(

\**sigh*\*

Reply:
Given that, do you agree that what I've done here is a reasonable solution (if I add a bit of explanation to tidy this up)? I guess the other approach would be to explicitly list each file into the `--files` arg.

Reply:
I think your solution is tidier than listing all files